pkg/tfgen/generate_csharp.go

```diff
@@ -0,0 +1,1312 @@
+	for i, word := range strings.Fields(comment) {
+		word = sanitizeForCSharpDocComment(word)
+		switch {
+		case i == 0:
+			word = fmt.Sprintf("%v/// %s", prefix, word)
+		case width+len(word)+1 > maxWidth:
+			w.WriteString("\n")
+			word = fmt.Sprintf("%v/// %s", prefix, word)
+			width = 0
+		default:
+			word = " " + word
+		}
+
+		w.WriteString(word)
+		width += len(word)
+	}
+	w.Writefmtln("")
+	w.Writefmtln("%s/// </summary>", prefix)
+}
+
+type csharpResourceGenerator struct {
+	g *csharpGenerator
+
+	mod    *module
+	res    *resourceType
+	fun    *resourceFunc
+	nested []*csharpNestedType
+
+	name string
+	w    *tools.GenWriter
+}
+
+func newCSharpResourceGenerator(g *csharpGenerator, mod *module, res *resourceType,
+	nested *csharpNestedTypes) *csharpResourceGenerator {
+
+	return &csharpResourceGenerator{
+		g:      g,
+		mod:    mod,
+		res:    res,
+		nested: nested.getDeclaredTypes(res),
+		name:   res.name,
+	}
+}
+
+func newCSharpDatasourceGenerator(g *csharpGenerator, mod *module, fun *resourceFunc,
+	nested *csharpNestedTypes) *csharpResourceGenerator {
+
+	return &csharpResourceGenerator{
+		g:      g,
+		mod:    mod,
+		fun:    fun,
+		nested: nested.getDeclaredTypes(fun),
+		name:   strings.Title(fun.name),
+	}
+}
+
+func (rg *csharpResourceGenerator) emit() (string, error) {
+	// Create a resource source file into which all of this resource's types will go.
+	// We need to check if the resource is called "utilities", as this may overlap with the utilities.cs file.
+	filename := rg.name
+	if filename == "Utilities" {
+		filename = fmt.Sprintf("%s_", filename)
+	}
+
+	w, file, err := rg.g.openWriter(rg.mod, filename+".cs")
+	if err != nil {
+		return "", err
+	}
+	defer contract.IgnoreClose(w)
+	rg.w = w
+
+	rg.openNamespace()
+	if rg.res != nil {
+		rg.generateResourceClass()
+		rg.generateResourceArgs()
+		rg.generateResourceState()
+	} else {
+		contract.Assert(rg.fun != nil)
+		rg.generateDatasourceFunc()
+		rg.generateDatasourceArgs()
+		rg.generateDatasourceResult()
+	}
+	rg.generateNestedTypes()
+	rg.closeNamespace()
+
+	return file, nil
+}
+
+func (rg *csharpResourceGenerator) openNamespace() {
+	rg.w.Writefmtln("#nullable enable")
+	rg.w.Writefmtln("")
+	rg.w.Writefmtln("using System.Collections.Immutable;")
+	rg.w.Writefmtln("using System.Threading.Tasks;")
+	rg.w.Writefmtln("using Pulumi.Serialization;")
+	rg.w.Writefmtln("")
+	rg.w.Writefmtln("namespace %s", rg.g.moduleNamespace(rg.mod))
+	rg.w.Writefmtln("{")
+}
+
+func (rg *csharpResourceGenerator) closeNamespace() {
+	rg.w.Writefmtln("}")
+}
+
+func (rg *csharpResourceGenerator) generateResourceClass() {
+	// Write the doc comment, if any.
+	if rg.res.doc != "" {
+		emitCSharpDocComment(rg.w, rg.res.doc, rg.res.docURL, "    ")
+	}
+
+	// Open the class.
+	className := rg.name
+	baseType := "Pulumi.CustomResource"
+	if rg.res.IsProvider() {
+		baseType = "Pulumi.ProviderResource"
+	}
+	rg.w.Writefmtln("    public class %s : %s", className, baseType)
```

Comment:
we should likely make resources into partial classes.  that way we can easily add overlays/mixins to them.